### DISCUSSION TITLE
Behaviour of association proxy "Simplifying Scalar Collections" example might be surprising to some people

### DISCUSSION THREAD

**Discussion by gordthompson** [sqlalchemy/sqlalchemy#11028] [Usage Questions] [answered]
I'm not sure why this question doesn't arise more often, especially on Stack Overflow, but I suspect that some users must find the behaviour surprising.

The example code for [Simplifying Scalar Collections](https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html#simplifying-scalar-collections) uses a many-to-many relationship between User and Keyword. One might expect that tagging multiple users with the same keyword (string) would associate each User with the same Keyword object, but that is not always the case. Using the exact code from the example cited above, and adding …

```python
if __name__ == "__main__":
    engine = create_engine("sqlite:///database.sqlite")

    Base.metadata.drop_all(engine)
    Base.metadata.create_all(engine)

    def dump_keywords():
        with engine.connect() as conn:
            print(
                f"keyword rows: {conn.exec_driver_sql('SELECT * FROM keyword').all()}"
            )

    with Session(engine) as sess:
        sess.add(gord := User(name="gord"))
        sess.add(zzzeek := User(name="zzzeek"))
        gord.keywords.append("nerd")
        sess.commit()
        dump_keywords()  # keyword rows: [(1, 'nerd')]
        zzzeek.keywords.append("nerd")
        sess.commit()
        dump_keywords()  # keyword rows: [(1, 'nerd'), (2, 'nerd')]
```

… shows that the same keyword (string) gets added to the "keyword" table more than once, hence there is more than one Keyword object with keyword="nerd".

My first thought would be to ditch the autoincrement keyword.id column and make keyword.keyword the PK, but lots and lots of people do something like this

```python
class Keyword(Base):
    __tablename__ = "keyword"
    id: Mapped[int] = mapped_column(primary_key=True)
    keyword: Mapped[str] = mapped_column(String(64), unique=True)
```

and that will fail when the code tries to add the second row for 'nerd' (unique constraint violation).

Is there a canonical response documented somewhere for this usage case?

**Comment by zzzeek** [comment, marked as answer]
the canonical response is the [UniqueObject](https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObject) recipe which is quite old and not particularly elegant since it uses an implicit thread local session.

There's no way to do this case without doing a database lookup, so if I were using this pattern myself, since I am a big proponent of the transational context being explicit, I'd make the `creator` here raise an error and only use the assocaition proxy for viewing.   manipulation of items would be via special methods that accept the `Session` so that it can look up an existing `Keyword`.

as these two approaches are very far apart, I dont consider either to be "canonical", and in this case the most I would say is add a note effectively saying "by the way, if you add the same string twice, you get two Keyword objects.  if you want the Keyword object to be unique on its name, the `creator` function would need adjustment to use an existing `Keyword` in the case of an already existing name.

as is usually the case here, when I look at this code, it seems obvious that the `Keyword` is created each time and there is no hash lookup or anything like that.
